### PR TITLE
feat(npu): llguidance bitmask path for Ascend NPU (#1107)

### PR DESCRIPTION
**Pull request by LongCatchen**
## Summary

Enable `--grammar-backend llguidance` on Ascend NPU by dispatching bitmask application to `torch.ops.npu.apply_token_bitmask` (#1107).

- NPU branch in `GuidanceGrammar.apply_vocab_mask`
- Mock-based unit tests in `test/registered/ascend/test_llguidance_npu_mask.py`

## Related

- https://github.com/Ascend/sglang/issues/1107

## Test plan

- [x] Unit tests (mock NPU ops, CPU CI compatible)
- [ ] Optional NPU smoke with structured output

---
### CI States

Latest PR Test (Base): :x: [Run #35582165950](https://github.com/sgl-project/sglang/actions/runs/35582165950)
Latest PR Test (Extra): :x: [Run #35582165617](https://github.com/sgl-project/sglang/actions/runs/35582165617)
Latest PR Test (AMD ROCm 10): :x: [Run #35582165737](https://github.com/sgl-project/sglang/actions/runs/35582165737)

### CODE REVIEW

**Comment by LongCatchen** [comment]
compile

**Comment by LongCatchen** [comment]
NPU local retest on Atlas 800T A2 (`sglang_dev`, NPU test host):

- **HEAD:** `f45f490`
- **Test:** `test/registered/ascend/test_llguidance_npu_mask.py`
- **Result:** 4 passed in ~14s

Ascend issue: https://github.com/Ascend/sglang/issues/1107

**Comment by gjsheu** [comment]
Please supply complete service startup and request invocation scripts, and show the execution results.

**Comment by LongCatchen** [comment]
Service startup and request scripts for `--grammar-backend llguidance` on Ascend NPU. The bitmask path is `GuidanceGrammar.apply_vocab_mask` -> `torch.ops.npu.apply_token_bitmask` when `logits.device.type == "npu"`.

### 1. Start the server

```bash
python -m sglang.launch_server \
  --model-path /path/to/model \
  --host 0.0.0.0 \
  --port 30000 \
  --attention-backend ascend \
  --grammar-backend llguidance
```

### 2. Send a constrained request

```bash
curl -s http://127.0.0.1:30000/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d "{\"model\":\"/path/to/model\",\"temperature\":0,\"max_tokens\":64,\"messages\":[{\"role\":\"user\",\"content\":\"Return the capital of France as JSON with keys name and population.\"}],\"response_format\":{\"type\":\"json_schema\",\"json_schema\":{\"name\":\"capital\",\"schema\":{\"type\":\"object\",\"properties\":{\"name\":{\"type\":\"string\"},\"population\":{\"type\":\"integer\"}},\"required\":[\"name\",\"population\"]}}}}"
```

A valid result is a JSON object with both required keys, for example `{"name":"Paris","population":2100000}`.

### 3. Execution result already collected

CPU-compatible unit test of the NPU dispatch (kernel mocked so CI can run it):

```bash
export PYTHONPATH=python
python3 -m pytest test/registered/ascend/test_llguidance_npu_mask.py -q --tb=short
```

```
4 passed, 15 warnings in 14.00s
```

HEAD at that run: `f45f490`. A fresh end-to-end generate was not repeated on the current test host: the container does not have this branch checked out, and there is no small text checkpoint mounted for a short serve.

**Comment by LongCatchen** [comment]
End-to-end retest on an NPU test host, using this PR branch at `f45f490` (the checkout under test, not the image tree).

### Unit test

```bash
export PYTHONPATH=python
python3 -m pytest test/registered/ascend/test_llguidance_npu_mask.py -q --tb=line
```

```
4 passed, 15 warnings in 37.96s
```

### Server

```bash
python -m sglang.launch_server \
  --model-path Qwen/Qwen2.5-0.5B-Instruct \
  --host 127.0.0.1 \
  --port 30010 \
  --device npu \
  --attention-backend ascend \
  --grammar-backend llguidance \
  --mem-fraction-static 0.25 \
  --skip-server-warmup
```

Startup log includes `grammar_backend=llguidance`, `attention_backend=ascend`, and `The server is fired up and ready to roll!`.

### Constrained request

`POST /v1/chat/completions` with `temperature=0`, `max_tokens=64`, and `response_format.type=json_schema` requiring `name` (string) and `population` (integer). Prompt: "Return the capital of France as JSON with keys name and population."

Assistant content:

```json
{
  "name": "Paris",
  "population": 2184000
}
```

`finish_reason=stop`, `completion_tokens=23`. The server was stopped after this request.

**Comment by LongCatchen** [comment]
Follow-up NPU retest for Ascend/sglang #1107 acceptance (JSON Schema / regex / EBNF / structural_tag), on a **single NPU** (`/dev/davinci0` only, `ASCEND_RT_VISIBLE_DEVICES=0`), branch `f45f490`.

### HTTP constrained generate

| Constraint | Result |
|---|---|
| `response_format.json_schema` | `{"name":"Paris","population":2184000}` |
| `regex: (Paris\|London)` | `Paris` |
| `ebnf: root ::= "yes" \| "no"` | `yes` |
| `structural_tag` (legacy structures/triggers) | Grammar compiles with real tokenizer special tokens (`<\|box_start\|>` / `<\|box_end\|>`). Free text before the trigger is allowed by the structural_tag grammar (`at_least_one=false`). |

### structural_tag NPU bitmask (after trigger)

After `accept_token(<|box_start|>)`, `GuidanceGrammar.apply_vocab_mask` on NPU logits keeps **7** allowed tokens, all `{` / `{\n` / `{"` variants — i.e. the JSON object start required by the schema. This exercises `torch.ops.npu.apply_token_bitmask`.

### Unit test

```
4 passed, 15 warnings in 37.96s
```

(`test/registered/ascend/test_llguidance_npu_mask.py`)

**Comment by LongCatchen** [comment]
### Follow-up after DeepSeek Pro/Flash review

Pushed to tip:
- `48ba0ae` — NPU mock install returns a `restore()` to undo `torch.ops.npu` / `torch.npu.is_available` patches after each test

Next: re-run ascend UT + a real NPU `apply_token_bitmask` smoke on the single-card container (`davinci0` only) and report results here.

**Comment by LongCatchen** [comment]
### Follow-up retest (tip `48ba0ae`)

Single-NPU only: container devices `/dev/davinci0` (+ manager/hdc), `ASCEND_RT_VISIBLE_DEVICES=0`.

**Unit test**
```
python3 -m pytest test/registered/ascend/test_llguidance_npu_mask.py -q --tb=short
4 passed, 15 warnings in 30.03s
```

**Real NPU kernel smoke** (`torch.ops.npu.apply_token_bitmask` on `npu:0`, not mocked)
- `npu_available True`, `device_count 1`
- `sgl_kernel_npu` import OK
- Allowed ids stay finite; blocked ids become `-inf`
- Result: **SMOKE_PASS**

This addresses the review concern that mock UT alone does not prove the real kernel signature/semantics.
